MM-12233: local user mention results

lieut-data opened the pull request:
#### Summary
Leverage existing data in the Redux store to present local at mentions even before the server responds with results. This is not strictly dependent on https://github.com/mattermost/mattermost-redux/pull/693, but that PR pre-fills the data using knowledge from the posts.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-12233

#### Checklist
- [x] Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit and component tests passed
- [x] Added or updated unit tests (required for all new features)
- [ ] Needs to be implemented in mobile (link to PR or User Story)
- [ ] Has server changes (please link)
- [x] Has redux changes: https://github.com/mattermost/mattermost-redux/pull/715
- [ ] Has UI changes
- [ ] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/mattermost-webapp/blob/master/i18n/en.json)) updates
- [ ] Touches critical sections of the codebase (auth, posting, etc.)

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-03ebd1a156c8edbd3.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-03ebd1a156c8edbd3

mattermod commented:
Spinmint test running for more than 7 days. This test server was terminated.

jasonblais commented:
@lieut-data Should I test this PR after the redux one is merged, or before? Should I also test the Redux PR separately?

lieut-data commented:
@jasonblais, this PR has the Redux commit baked into it, so it's worth testing together. I've added you as reviewer to https://github.com/mattermost/mattermost-redux/pull/693 explicitly to get your feedback on that part of the change.

jasonblais commented:
Thanks Jesse, just saw the note in the redux PR. Will review both 👍

lieut-data commented:
Heads up that there is an issue with the with the RHS throwing an exception for which I'll push up a commit for, but I won't bother recreating the test server right away.

mattermod commented:
Spinmint test server created at: https://i-09ee185aed4ba4f79.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-09ee185aed4ba4f79

lieut-data commented:
Thanks for the feedback, @jasonblais!

1. Interesting: I'll have to dig deeper on this one.

2. - 3. I'm glad you observed this: the generated data data on the spinmint servers isn't realistic, as posts are made from users at random without taking into account if they are in the channel. This dissonance helps illustrate the effect of the Redux commit, though it does make it hard to assess the "correctness" of this implementation.

4. I agree that the potential for confusion is present. We could certainly drop the Redux commit from the equation, at the cost of no longer having immediate results on first load. I'm wondering if, instead, we might rebrand the autocomplete from `Channel Members` and `Not In Channel` to `Recent Users` and `Other Users`, effectively making the sections more ambiguous. Thoughts?

jasonblais commented:
4 - There might certainly be UX changes we could consider - some competitors don't separate users in the autocomplete at all.

To confirm:
- without Redux commit, we present local at mention suggestions immediately, followed by the server results
- with Redux commit, we do the same but ALSO have them on first load.

Is the above correct?

5 - Question: Do we need to make any changes on mobile?

mattermod commented:
Spinmint test running for more than 7 days. This test server was terminated.

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-0dd0e90f4d7ec00bf.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-0dd0e90f4d7ec00bf

lieut-data commented:
4 - As previously discussed, we've discarded the original Redux PR that prepopulates the data structure with post data, and instead replaced it one that maintains these data structured on user added/removed websocket events.

5, @enahum - Yes, we should investigate doing this on mobile as well. I've filed https://mattermost.atlassian.net/browse/MM-13076 to track this.

lieut-data commented:
@jasonblais, I've updated the spinmint with the latest changes, as discussed earlier last week. (Note that the "fake data" problem is still present.)

jasonblais commented:
Looks like 1 is not yet fixed for me on the spinmint.

2 -5 👍 

6 - Another issue I noticed is that "Special Mentions" now always appear first on the list.

![image](https://user-images.githubusercontent.com/13119842/48742631-c1912b00-ec2d-11e8-9a49-de7097bbbb95.png)

Previously it was that 'channel members' were listed first, then 'special mentions':
![image](https://user-images.githubusercontent.com/13119842/48742618-b3430f00-ec2d-11e8-8d6b-370333501ed3.png)

----

Otherwise this looks good to me, though testing on a realistic environment (i.e. community.mattermost.com) will be the real test. 

Given the large changes here, would you feel that QA smoke testing prior to merging is advisable?

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

lieut-data commented:
@jasonblais,

1 - Thanks for the reminder! I had thought this was fixed when I rebased from master, but in reality, my reproduction steps were off. I can reproduce this 100% -- on this spinmint and on pre-release -- if I type `@` in a channel (see the suggestion box), then switch to another channel and type `@` again. The textbox fails to reset the pretext between channel changes, and thinks nothing has changed, so it doesn't trigger the matching. I'll see if I can fix this one as part of my changes, but given it reproduces on `master`, I'm inclined to make it non-blocking relative to this PR.

6 - Good eye! I've changed the order here.

100% to get QA involved here for a smoke test. @lindy65, would you be willing and available to help qualify this functionality?

mattermod commented:
Spinmint test server created at: https://i-0a33284ea1a391a58.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-0a33284ea1a391a58

jasonblais commented:
1 - Hm, I wasn't able to reproduce on `master`. But if you're able to, definitely fine to create a ticket and fix it separately.

Else LGTM 👍

lieut-data commented:
Sounds good: I've filed https://mattermost.atlassian.net/browse/MM-13085 to track the "@" mention issue.

lindy65 commented:
Thanks @lieut-data @jasonblais 

Tested along the same lines as Jason

1. Confirm I was able to reproduce on spinmint and master - ticket created by Jesse - https://mattermost.atlassian.net/browse/MM-13085

2 - 5. 👍 

6. Fixed - (`channel members` listed first, then `special mentions` then `not in channel`)

Did some additional smoke testing around user mentions:

I get a blank screen when trying to edit a channel header - does _not_ reproduce on master. (I was wanting to add a user mention in a channel header).  This occurs for public, private, DM and GM channels. The other options in the channel dropdown menu work as expected. Also get a blank screen when trying to add a header via the link at the top of a new channel. (Not sure if related to this PR)

![image](https://user-images.githubusercontent.com/13020833/48758861-1a060e00-ecaa-11e8-841e-e5efaafefb52.png)

Otherwise looks good to me :)

jasonblais commented:
Nice catch Lindy! Here's the associated JavaScript error

![image](https://user-images.githubusercontent.com/13119842/48773255-d48b1600-ec94-11e8-87ea-ba1a8faaf029.png)

Wondering if it's just an outdated Redux commit in this PR branch that's leading to the error?

lindy65 commented:
Thanks for including the JavaScript error @jasonblais 👍

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-0dd89fd27c64fa7b2.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-0dd89fd27c64fa7b2

lieut-data commented:
Thanks, @lindy65 & @jasonblais -- fixed the channel header, and inspected the remaining code paths to make sure this was an isolated regression.

lindy65 commented:
Thanks @lieut-data 

Confirm that editing the channel header is fixed in public, private, DM and GM channels.

Also smoke tested along similar lines as before and apart from issues already covered and ticket opened to address the `@` mention issue, I did not find any additional issues.

Looks good to me.

mattermod commented:
Spinmint test server destroyed

jasonblais commented:
Thanks all! @lieut-data I moved this to dev review, please help queue it for the appropriate devs :)

lieut-data commented on components/create_comment/create_comment.jsx:
```diff
@@ -444,7 +444,9 @@ export default class CreateComment extends React.PureComponent {
             e.persist();
-            this.refs.textbox.blur();
+            if (this.refs.textbox) {
+                this.refs.textbox.getWrappedInstance().blur();
```
I'd love feedback on connecting `Textbox` and updating the world to poke through the wrapped instance like this. An alternative was to push up the the store access to all components that used the `Textbox`, but that seems like a violation of concerns. Alternatively, I could have accessed the store globally from within the suggestion provider, but there's currently no way to "trigger" the user of the provider to try a pretext match.

hmhealey replied:
I'm not sure I understand. Do you just mean to avoid the mess of `getWrappedInstance`?

lieut-data replied:
Yeah, that :/

lieut-data commented on components/textbox/textbox.jsx:
```diff
@@ -52,7 +58,12 @@ export default class Textbox extends React.Component {
 
         this.suggestionProviders = [
-            new AtMentionProvider(this.props.channelId),
+            new AtMentionProvider({
```
Ideally, `AtMentionProvider` would be connected itself, but while this is technically possible, it still leaves us without a way to back-propagate pretext changes on props-only changes unless `Textbox` is in on the loop.

hmhealey commented on components/suggestion/at_mention_provider/at_mention_provider.jsx:
```diff
@@ -0,0 +1,190 @@
+    }
+
+    // setProps gives the provider additional context for matching pretexts. Ideally this would
+    // just be something akin to a connected component with access to the store itself.
```
This is something I've wanted to look at as well. We could either:
1. Change the Provider classes to components that just render nothing so that we can render something like `<SuggestionBox><AtMentionProvider/></SuggestionBox>`
2. Hook into the context when instantiating the `Textbox` to pass the store directly to the Provider

The first method is a bit hacky since the provider is not really a component, but it is cleaner than the second one since that requires us to breach the encapsulation provided by `react-redux`

lieut-data replied:
The first also naturally moves us towards triggering a "render" (i.e. prefix match) on any change, while the latter has the difficulty of still needing to notify the renderer that the provider has changed, perhaps outside of an explicit pretext match. Complicated either way, I suppose!

hmhealey commented on components/suggestion/suggestion_list.jsx:
```diff
@@ -65,6 +65,10 @@ export default class SuggestionList extends React.PureComponent {
             const contentBottomPadding = parseInt(content.css('padding-top'), 10);
 
             const item = $(ReactDOM.findDOMNode(this.refs[term]));
+            if (!item[0]) {
```
Minor nit, but could this be `item.length === 0`? It might just be a bit clearer

hmhealey commented on components/textbox/index.js:
```diff
@@ -0,0 +1,34 @@
+
+import Textbox from './textbox.jsx';
+
+const autocompleteUsersInChannel = (prefix) => (dispatch, getState) => {
```
This should probably live in `actions/views`

hmhealey commented on utils/constants.jsx:
```diff
@@ -1269,6 +1269,7 @@ export const Constants = {
     MENTION_MORE_CHANNELS: 'mention.morechannels',
     MENTION_UNREAD_CHANNELS: 'mention.unread.channels',
     MENTION_MEMBERS: 'mention.members',
+    MENTION_MORE_MEMBERS: 'mention.moremembers',
```
Does this need to be wrapped in the `t` function somewhere?

hmhealey commented on i18n/en.json:
```diff
@@ -2567,6 +2567,7 @@
   "suggestion.mention.channels": "My Channels",
   "suggestion.mention.here": "Notifies everyone in the channel and online",
   "suggestion.mention.members": "Channel Members",
+  "suggestion.mention.moremembers": "Other Members",
```
Should this be "Other Users"? Without "Channel" or "Team" for context, "Members" looks weird to be

jasonblais replied:
Where/when does this string occur?

I feel it was discussed, but I forget now, and didn't spot it on the latest spinmint below.

lieut-data replied:
This would appear over the loading section, but disappears once everything is merged together.

![image](https://user-images.githubusercontent.com/1023171/48878651-5d5d9b00-edd6-11e8-89b3-d7beece3d09e.png)

jasonblais replied:
Got it. I think `Members` referring to channel members is quite clear in this context.

hmhealey replied:
Are those users channel members though? I thought this was for loading people who aren't in the channel

jasonblais replied:
@hmhealey Some are channel members, some are not.

We are first loading users locally, so the query returns immediate results. We then load everyone else, most of which are not in the channel but a few that could be.

Once the loading has finished, there will be three sections: Channel Members, Special Mentions, and Not In Channel

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-08a63d874aaf7e2bc.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-08a63d874aaf7e2bc

jasonblais commented:
@crspeller @hmhealey this one is ready for re-review

mattermod commented:
Spinmint test server destroyed

mattermod commented:
Spinmint test running for more than 7 days. This test server was terminated.

mattermod commented:
Spinmint test running for more than 7 days. This test server was terminated.
